packages/core/src/operations/ensure.ts

```diff
@@ -1,22 +1,23 @@
-import { Operation, Resource } from '../operation';
-import { spawn } from './spawn';
+import { Operation } from '../operation';
+import { createFuture } from '../future';
+import { withLabels } from '../labels';
 
-
-export function ensure<T>(fn: () => Operation<T> | void): Resource<undefined> {
-  return {
-    name: 'ensure',
-    *init() {
-      yield spawn(function*() {
-        try {
-          yield;
-        } finally {
-          let result = fn();
-          if(result) {
-            yield result;
-          }
-        }
-      }, { labels: { name: 'ensureHandler' } });
-      return undefined;
+export function ensure<T>(fn: () => Operation<T> | void): Operation<undefined> {
+  return withLabels((task) => {
+    let { scope } = task.options;
+    if(!scope) {
+      throw new Error('cannot run `ensure` on a task without scope');
     }
-  };
+    scope.run(function*() {
+      try {
+        yield;
+      } finally {
+        let result = fn();
+        if(result) {
+          yield result;
+        }
+      }
+    }, { labels: { name: 'ensureHandler' } });
+    return createFuture({ state: 'completed', value: undefined }).future;
+  }, { name: 'ensure' });
```

Comment:
We can avoid all the labelling just by using functions with names, right?

Reply:
Yes, I think so, good point!